runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -905,6 +928,19 @@ public final void notifyCheckpointComplete(long checkpointId) throws Exception {
       bufferingDoFnRunner.checkpointCompleted(checkpointId);
     }
 
+    List<InMemoryBundleFinalizer.Finalization> finalizations =
+        pendingFinalizations.get(checkpointId);
+    if (finalizations != null) {
+      // remove old finalizations except for the current one
+      pendingFinalizations.clear();
+      pendingFinalizations.put(checkpointId, finalizations);
+
+      // confirm all finalizations
+      for (InMemoryBundleFinalizer.Finalization finalization : finalizations) {
+        finalization.getCallback().onBundleSuccess();
```

Comment:
Acknowledging the latest bundle finalize will also acknowledge all earlier requests for bundle finalization?

Reply:
Between two checkpoint boundaries, the DoFn may have requested multiple finalizations (e.g. one per element processed) each having a different side effect.